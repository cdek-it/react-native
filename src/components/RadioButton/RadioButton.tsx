import React, { memo, useCallback, useState } from 'react'
import { Pressable, View } from 'react-native'
import Animated, { LinearTransition } from 'react-native-reanimated'

import { makeStyles } from '../../utils/makeStyles'

import { useStateStyles } from './hooks/useStateStyles'

export interface RadioButtonProps {
  /** Обработчик нажатия на кнопку */
  onPress: () => void
  /**
   * true, если необходим компонент в активном состоянии
   * @default false
   */
  checked?: boolean
  /**
   * Управление доступностью компонента
   * @default false
   */
  disabled?: boolean
  /** Выбор состояния компонента */
  state?: 'default' | 'danger'
}

export const RadioButton = memo<RadioButtonProps>(({ onPress, checked, disabled, state }) => {
  const styles = useStyles()
  const [pressed, setPressed] = useState(false)
  const stateStyles = useStateStyles(checked, disabled, pressed, state)

  const onPressIn = useCallback(() => setPressed(true), [])
  const onPressOut = useCallback(() => setPressed(false), [])

  return (
    <>
      {!disabled && state === 'danger' && (
        <Animated.View layout={LinearTransition.duration(100)} style={[styles.outline]} />
      )}
      <Pressable
        disabled={disabled}
        style={[styles.container, stateStyles.viewState]}
        testID='RadioButton_Pressable'
        onPress={onPress}
        onPressIn={onPressIn}
        onPressOut={onPressOut}
      >
        <View style={[styles.center, stateStyles.centerViewBackground]} />
      </Pressable>
    </>
  )
})

const useStyles = makeStyles(({ theme }) => ({
  container: {
    position: 'absolute',
    width: theme.Form.RadioButton.radiobuttonWidth,
    height: theme.Form.RadioButton.radiobuttonHeight,
    borderRadius: theme.Form.RadioButton.radiobuttonWidth,
    alignItems: 'center',
    justifyContent: 'center',
    marginLeft: theme.General.focusShadowWidth,
    marginTop: theme.General.focusShadowWidth,
  },
  center: {
    width: theme.Form.RadioButton.radiobuttonIconSize,
    height: theme.Form.RadioButton.radiobuttonIconSize,
    borderRadius: theme.Form.RadioButton.radiobuttonIconSize,
  },
  outline: {
    position: 'absolute',
    width: theme.Form.RadioButton.radiobuttonWidth + theme.General.focusShadowWidth * 2,
    height: theme.Form.RadioButton.radiobuttonHeight + theme.General.focusShadowWidth * 2,
    borderRadius: theme.Form.RadioButton.radiobuttonHeight + theme.General.focusShadowWidth * 2,
    backgroundColor: theme.General.focusOutlineErrorColor,
  },
}))
