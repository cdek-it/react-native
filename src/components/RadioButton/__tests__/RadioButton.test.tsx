import { render, userEvent } from '@testing-library/react-native'
import React from 'react'

import { RadioButton, type RadioButtonProps } from '../RadioButton'

describe('RadioButton', () => {
  const defaultProps: RadioButtonProps = {
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    onPress: () => {},
  }

  describe('snapshots', () => {
    const snapshotCases: Array<[string, Partial<RadioButtonProps>]> = [
      ['default', {}],
      [
        'checked = true, disabled = false, state = danger',
        {
          checked: true,
          disabled: false,
          state: 'danger',
        },
      ],
      [
        'checked = false, disabled = false, state = danger',
        {
          checked: false,
          disabled: false,
          state: 'danger',
        },
      ],

      [
        'checked = true, disabled = false, state = default',
        {
          checked: true,
          disabled: false,
          state: 'default',
        },
      ],
      [
        'checked = false, disabled = false, state = default',
        {
          checked: false,
          disabled: false,
          state: 'default',
        },
      ],

      [
        'checked = true, disabled = true, state = default',
        {
          checked: true,
          disabled: true,
          state: 'default',
        },
      ],
      [
        'checked = false, disabled = true, state = default',
        {
          checked: false,
          disabled: true,
          state: 'default',
        },
      ],
    ]

    beforeAll(() => {
      jest.mock('react-native/Libraries/Components/Pressable/Pressable', () => ({
        default: 'Pressable',
      }))
    })

    afterAll(() => {
      jest.unmock('react-native/Libraries/Components/Pressable/Pressable')
    })

    test.each(snapshotCases)('%s', (_, props) => {
      const { toJSON } = render(<RadioButton {...defaultProps} {...props} />)
      expect(toJSON()).toMatchSnapshot()
    })
  })

  test('should handle press', async () => {
    const mockedOnPress = jest.fn()
    const { queryByTestId } = render(<RadioButton onPress={mockedOnPress} />)
    const pressable = queryByTestId('RadioButton_Pressable')
    const user = userEvent.setup()

    await user.press(pressable)
    expect(mockedOnPress).toHaveBeenCalled()
  })

  test('should NOT handle press', async () => {
    const mockedOnPress = jest.fn()
    const { queryByTestId } = render(<RadioButton disabled onPress={mockedOnPress} />)
    const pressable = queryByTestId('RadioButton_Pressable')
    const user = userEvent.setup()

    await user.press(pressable)
    expect(mockedOnPress).not.toHaveBeenCalled()
  })
})
