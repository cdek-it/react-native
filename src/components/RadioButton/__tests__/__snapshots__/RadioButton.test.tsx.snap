// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton snapshots checked = false, disabled = false, state = danger 1`] = `
[
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    jestInlineStyle={
      [
        {
          "backgroundColor": "#f6d4d1",
          "borderRadius": 28,
          "height": 28,
          "position": "absolute",
          "width": 28,
        },
      ]
    }
    layout={
      LinearTransition {
        "build": [Function],
        "durationV": 100,
        "randomizeDelay": false,
        "reduceMotionV": "system",
      }
    }
    style={
      {
        "backgroundColor": "#f6d4d1",
        "borderRadius": 28,
        "height": 28,
        "position": "absolute",
        "width": 28,
      }
    }
  />,
  <Pressable
    disabled={false}
    onPress={[Function]}
    onPressIn={[Function]}
    onPressOut={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 21,
          "height": 21,
          "justifyContent": "center",
          "marginLeft": 3.5,
          "marginTop": 3.5,
          "position": "absolute",
          "width": 21,
        },
        [
          {
            "backgroundColor": "#ffffff",
            "borderColor": "#dbdada",
            "borderWidth": 1,
          },
          false,
          false,
          false,
          {
            "borderColor": "#e26e63",
            "borderWidth": 1,
          },
          false,
          false,
          false,
          false,
        ],
      ]
    }
    testID="RadioButton_Pressable"
  >
    <View
      style={
        [
          {
            "borderRadius": 10.5,
            "height": 10.5,
            "width": 10.5,
          },
          [
            {
              "backgroundColor": "#ffffff",
            },
            false,
          ],
        ]
      }
    />
  </Pressable>,
]
`;

exports[`RadioButton snapshots checked = false, disabled = false, state = default 1`] = `
<Pressable
  disabled={false}
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 21,
        "height": 21,
        "justifyContent": "center",
        "marginLeft": 3.5,
        "marginTop": 3.5,
        "position": "absolute",
        "width": 21,
      },
      [
        {
          "backgroundColor": "#ffffff",
          "borderColor": "#dbdada",
          "borderWidth": 1,
        },
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
      ],
    ]
  }
  testID="RadioButton_Pressable"
>
  <View
    style={
      [
        {
          "borderRadius": 10.5,
          "height": 10.5,
          "width": 10.5,
        },
        [
          {
            "backgroundColor": "#ffffff",
          },
          false,
        ],
      ]
    }
  />
</Pressable>
`;

exports[`RadioButton snapshots checked = false, disabled = true, state = default 1`] = `
<Pressable
  disabled={true}
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 21,
        "height": 21,
        "justifyContent": "center",
        "marginLeft": 3.5,
        "marginTop": 3.5,
        "position": "absolute",
        "width": 21,
      },
      [
        {
          "backgroundColor": "#ffffff",
          "borderColor": "#dbdada",
          "borderWidth": 1,
        },
        false,
        false,
        false,
        false,
        false,
        false,
        {
          "backgroundColor": "rgba(0, 0, 0, 0.1000)",
          "borderColor": "#dbdada",
          "borderWidth": 1,
          "opacity": 0.6,
        },
        false,
      ],
    ]
  }
  testID="RadioButton_Pressable"
>
  <View
    style={
      [
        {
          "borderRadius": 10.5,
          "height": 10.5,
          "width": 10.5,
        },
        [
          {
            "backgroundColor": "#ffffff",
          },
          {
            "backgroundColor": "transparent",
          },
        ],
      ]
    }
  />
</Pressable>
`;

exports[`RadioButton snapshots checked = true, disabled = false, state = danger 1`] = `
[
  <View
    collapsable={false}
    jestAnimatedStyle={
      {
        "value": {},
      }
    }
    jestInlineStyle={
      [
        {
          "backgroundColor": "#f6d4d1",
          "borderRadius": 28,
          "height": 28,
          "position": "absolute",
          "width": 28,
        },
      ]
    }
    layout={
      LinearTransition {
        "build": [Function],
        "durationV": 100,
        "randomizeDelay": false,
        "reduceMotionV": "system",
      }
    }
    style={
      {
        "backgroundColor": "#f6d4d1",
        "borderRadius": 28,
        "height": 28,
        "position": "absolute",
        "width": 28,
      }
    }
  />,
  <Pressable
    disabled={false}
    onPress={[Function]}
    onPressIn={[Function]}
    onPressOut={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "borderRadius": 21,
          "height": 21,
          "justifyContent": "center",
          "marginLeft": 3.5,
          "marginTop": 3.5,
          "position": "absolute",
          "width": 21,
        },
        [
          {
            "backgroundColor": "#ffffff",
            "borderColor": "#dbdada",
            "borderWidth": 1,
          },
          false,
          {
            "backgroundColor": "#188700",
            "borderColor": "#188700",
            "borderWidth": 5,
          },
          false,
          {
            "borderColor": "#e26e63",
            "borderWidth": 1,
          },
          {
            "backgroundColor": "#188700",
            "borderColor": "#e26e63",
            "borderWidth": 1,
          },
          false,
          false,
          false,
        ],
      ]
    }
    testID="RadioButton_Pressable"
  >
    <View
      style={
        [
          {
            "borderRadius": 10.5,
            "height": 10.5,
            "width": 10.5,
          },
          [
            {
              "backgroundColor": "#ffffff",
            },
            false,
          ],
        ]
      }
    />
  </Pressable>,
]
`;

exports[`RadioButton snapshots checked = true, disabled = false, state = default 1`] = `
<Pressable
  disabled={false}
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 21,
        "height": 21,
        "justifyContent": "center",
        "marginLeft": 3.5,
        "marginTop": 3.5,
        "position": "absolute",
        "width": 21,
      },
      [
        {
          "backgroundColor": "#ffffff",
          "borderColor": "#dbdada",
          "borderWidth": 1,
        },
        false,
        {
          "backgroundColor": "#188700",
          "borderColor": "#188700",
          "borderWidth": 5,
        },
        false,
        false,
        false,
        false,
        false,
        false,
      ],
    ]
  }
  testID="RadioButton_Pressable"
>
  <View
    style={
      [
        {
          "borderRadius": 10.5,
          "height": 10.5,
          "width": 10.5,
        },
        [
          {
            "backgroundColor": "#ffffff",
          },
          false,
        ],
      ]
    }
  />
</Pressable>
`;

exports[`RadioButton snapshots checked = true, disabled = true, state = default 1`] = `
<Pressable
  disabled={true}
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 21,
        "height": 21,
        "justifyContent": "center",
        "marginLeft": 3.5,
        "marginTop": 3.5,
        "position": "absolute",
        "width": 21,
      },
      [
        {
          "backgroundColor": "#ffffff",
          "borderColor": "#dbdada",
          "borderWidth": 1,
        },
        false,
        {
          "backgroundColor": "#188700",
          "borderColor": "#188700",
          "borderWidth": 5,
        },
        false,
        false,
        false,
        false,
        {
          "backgroundColor": "rgba(0, 0, 0, 0.1000)",
          "borderColor": "#dbdada",
          "borderWidth": 1,
          "opacity": 0.6,
        },
        {
          "backgroundColor": "grey",
          "borderColor": "grey",
          "borderWidth": 5,
        },
      ],
    ]
  }
  testID="RadioButton_Pressable"
>
  <View
    style={
      [
        {
          "borderRadius": 10.5,
          "height": 10.5,
          "width": 10.5,
        },
        [
          {
            "backgroundColor": "#ffffff",
          },
          false,
        ],
      ]
    }
  />
</Pressable>
`;

exports[`RadioButton snapshots default 1`] = `
<Pressable
  onPress={[Function]}
  onPressIn={[Function]}
  onPressOut={[Function]}
  style={
    [
      {
        "alignItems": "center",
        "borderRadius": 21,
        "height": 21,
        "justifyContent": "center",
        "marginLeft": 3.5,
        "marginTop": 3.5,
        "position": "absolute",
        "width": 21,
      },
      [
        {
          "backgroundColor": "#ffffff",
          "borderColor": "#dbdada",
          "borderWidth": 1,
        },
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
      ],
    ]
  }
  testID="RadioButton_Pressable"
>
  <View
    style={
      [
        {
          "borderRadius": 10.5,
          "height": 10.5,
          "width": 10.5,
        },
        [
          {
            "backgroundColor": "#ffffff",
          },
          false,
        ],
      ]
    }
  />
</Pressable>
`;
